### ISSUE TITLE
Wrong type passed to withFieldConst results with ClassCastException

### ISSUE THREAD

**Issue by krzemin** [scalalandio/chimney#66]
Minimal example:

```
case class Foo1(y: String)
case class Foo2(y: String, x: Int)

Foo1("test")
    .into[Foo2]
    .withFieldConst(_.x, "xyz")
    .transform
```

Unfortunately, this one results with `ClassCastException` at runtime :(

The problem is caused by the fact that both selector and passed value refers to the same type variable, which in this case is inferred to `Any`. When type is passed explicitly, the code is not compiling, as expected. But I think we can do better that just leaving advice to always pass types explicitly :)

The idea for fix would be to modify signature of `withFieldConst` by introducing separate type variable:
`def withFieldConst[T, U](selector: To => T, value: U): TransformerInto[From, To, _]`

Then we would like to verify that `U` is proper subtype of `T`. We can't do this by type bound (`U <: T`) as this would be still inferred in a single step to `Any`.

We could introduce something similar to type-level evidence `implicit ev: U <:< T`, but most probably inside macro code, as this is whitebox macro that already takes implicit weak type tags. Plus: in macro we can format pretty error message.

Note that most probably the same problem may exists in `withFieldComputed` and `withFieldRenamed`. Hopefully, idea for fix would be just the same.

**Comment by kubukoz** [comment]
Would this work? `def withFieldConst[T, U](selector: To => T, value: U)(implicit ev: T =:= U)`

**Comment by krzemin** [comment]
Well, almost.
We require only subtyping relation between U and T. The function is implemented with whitebox macro that already takes implicit weak type tags, so just putting such evidence would interfere with macro signature.
The same subtyping check we can do in macro code, which I think would work well.

**Comment by kubukoz** [comment]
Sounds like a plan

**Comment by krzemin** [comment]
Would you like to try it out?

**Comment by kubukoz** [comment]
I don't have a usecase yet, I guess. Just added chimney to my project :P

**Comment by krzemin** [comment]
I meant fixing the issue :p

**Comment by kubukoz** [comment]
Oh, I wouldn't have time for it now anyway :(

**Comment by krzemin** [comment]
Closing as fixed with #68
